### PR TITLE
Implement deterministics

### PR DESCRIPTION
**Pull request by lucianopaz**
This PR builds on #147 but goes beyond it to implement deterministics that are traced while sampling. It does so by building a callback function to compute the deterministic after each kernel step.

I added a few tests, but XLA sampling does not work. So this PR can be a starting point for deterministics in pymc4 to build upon.

### CODE REVIEW

**Comment by junpenglao** [comment]
What type of error you are seeing under XLA?

**Comment by lucianopaz** [comment]
An `AttributeError`

test_sampling.py:39: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../pymc4/inference/sampling.py:139: in sample
    run_chains, inputs=[init_state, step_size]
../../../anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/compiler/xla/xla.py:108: in compile
    return xla_compile_wrapper()
../../../anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py:457: in __call__
    result = self._call(*args, **kwds)
../../../anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py:503: in _call
    self._initialize(args, kwds, add_initializers_to=initializer_map)
../../../anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py:408: in _initialize
    *args, **kwds))
../../../anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py:1848: in _get_concrete_function_internal_garbage_collected
    graph_function, _, _ = self._maybe_define_function(args, kwargs)
../../../anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py:2150: in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
../../../anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py:2041: in _create_graph_function
    capture_by_value=self._capture_by_value),
../../../anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/framework/func_graph.py:915: in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
../../../anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py:358: in wrapped_fn
    return weak_wrapped_fn().__wrapped__(*args, **kwds)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (), kwargs = {}

    def wrapper(*args, **kwargs):
      """Calls a converted version of original_func."""
      # TODO(mdan): Push this block higher in tf.function's call stack.
      try:
        return autograph.converted_call(
            original_func,
            autograph.ConversionOptions(
                recursive=True,
                optional_features=autograph_options,
                user_requested=True,
            ), args, kwargs)
      except Exception as e:  # pylint:disable=broad-except
        if hasattr(e, "ag_error_metadata"):
>         raise e.ag_error_metadata.to_exception(e)
E         AttributeError: in converted code:
E             relative to /home/lpaz:
E         
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/compiler/xla/xla.py:106 xla_compile_wrapper  *
E                 return _compile_internal(computation, inputs)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/compiler/xla/xla.py:352 _compile_internal
E                 outputs = computation(*computation_inputs)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py:457 __call__
E                 result = self._call(*args, **kwds)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py:503 _call
E                 self._initialize(args, kwds, add_initializers_to=initializer_map)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py:408 _initialize
E                 *args, **kwds))
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py:1848 _get_concrete_function_internal_garbage_collected
E                 graph_function, _, _ = self._maybe_define_function(args, kwargs)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py:2150 _maybe_define_function
E                 graph_function = self._create_graph_function(args, kwargs)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/eager/function.py:2041 _create_graph_function
E                 capture_by_value=self._capture_by_value),
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/framework/func_graph.py:915 func_graph_from_py_func
E                 func_outputs = python_func(*func_args, **func_kwargs)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/eager/def_function.py:358 wrapped_fn
E                 return weak_wrapped_fn().__wrapped__(*args, **kwds)
E             repos/pymc4/pymc4/inference/sampling.py:132 run_chains
E                 **(sample_chain_kwargs or dict()),
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/sample.py:362 sample_chain
E                 parallel_iterations=parallel_iterations)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/internal/util.py:371 trace_scan
E                 parallel_iterations=parallel_iterations)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/control_flow_ops.py:2478 while_loop_v2
E                 return_same_structure=True)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/control_flow_ops.py:2675 while_loop
E                 back_prop=back_prop)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/while_v2.py:198 while_loop
E                 add_control_dependencies=add_control_dependencies)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/framework/func_graph.py:915 func_graph_from_py_func
E                 func_outputs = python_func(*func_args, **func_kwargs)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/while_v2.py:176 wrapped_body
E                 outputs = body(*_pack_sequence_as(orig_loop_vars, args))
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/internal/util.py:360 _body
E                 state = loop_fn(state, elems_array.read(i))
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/sample.py:346 _trace_scan_fn
E                 parallel_iterations=parallel_iterations)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/internal/util.py:287 smart_for_loop
E                 parallel_iterations=parallel_iterations
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/control_flow_ops.py:2478 while_loop_v2
E                 return_same_structure=True)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/control_flow_ops.py:2675 while_loop
E                 back_prop=back_prop)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/while_v2.py:198 while_loop
E                 add_control_dependencies=add_control_dependencies)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/framework/func_graph.py:915 func_graph_from_py_func
E                 func_outputs = python_func(*func_args, **func_kwargs)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/while_v2.py:176 wrapped_body
E                 outputs = body(*_pack_sequence_as(orig_loop_vars, args))
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/internal/util.py:285 <lambda>
E                 body=lambda i, *args: [i + 1] + list(body_fn(*args)),
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/dual_averaging_step_size_adaptation.py:460 one_step
E                 current_state, inner_results)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/nuts.py:370 one_step
E                 parallel_iterations=self.parallel_iterations,
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/control_flow_ops.py:2478 while_loop_v2
E                 return_same_structure=True)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/control_flow_ops.py:2675 while_loop
E                 back_prop=back_prop)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/while_v2.py:198 while_loop
E                 add_control_dependencies=add_control_dependencies)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/framework/func_graph.py:915 func_graph_from_py_func
E                 func_outputs = python_func(*func_args, **func_kwargs)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/while_v2.py:176 wrapped_body
E                 outputs = body(*_pack_sequence_as(orig_loop_vars, args))
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/nuts.py:365 <lambda>
E                 metastate),
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/nuts.py:539 loop_tree_doubling
E                 momentum_state_memory)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/nuts.py:720 _build_sub_tree
E                 parallel_iterations=self.parallel_iterations
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/control_flow_ops.py:2478 while_loop_v2
E                 return_same_structure=True)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/control_flow_ops.py:2675 while_loop
E                 back_prop=back_prop)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/while_v2.py:198 while_loop
E                 add_control_dependencies=add_control_dependencies)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/framework/func_graph.py:915 func_graph_from_py_func
E                 func_outputs = python_func(*func_args, **func_kwargs)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/while_v2.py:176 wrapped_body
E                 outputs = body(*_pack_sequence_as(orig_loop_vars, args))
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/nuts.py:708 <lambda>
E                 not_divergence, momentum_state_memory)),
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/nuts.py:757 _loop_build_sub_tree
E                 prev_tree_state.target_grad_parts)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/internal/leapfrog_integrator.py:289 __call__
E                 target_grad_parts,
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/control_flow_ops.py:2478 while_loop_v2
E                 return_same_structure=True)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/control_flow_ops.py:2675 while_loop
E                 back_prop=back_prop)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/while_v2.py:198 while_loop
E                 add_control_dependencies=add_control_dependencies)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/framework/func_graph.py:915 func_graph_from_py_func
E                 func_outputs = python_func(*func_args, **func_kwargs)
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/ops/while_v2.py:176 wrapped_body
E                 outputs = body(*_pack_sequence_as(orig_loop_vars, args))
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/internal/leapfrog_integrator.py:283 <lambda>
E                 body=lambda i, *args: [i + 1] + list(self._one_step(*args)),  # pylint: disable=no-value-for-parameter
E             anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_probability/python/mcmc/internal/leapfrog_integrator.py:332 _one_step
E                 ng.set_shape(g.shape)
E         
E             AttributeError: 'IndexedSlices' object has no attribute 'set_shape'

../../../anaconda3/envs/pymc4/lib/python3.7/site-packages/tensorflow_core/python/framework/func_graph.py:905: AttributeError

**Comment by lucianopaz** [comment]
Looking at the diffs from this PR I noticed some code that was used in the stateful caching of deterministics, and when `Deterministic` was a `Distribution` subclass, but is useless now. I'm pushing a couple of commits to clean this up a bit.

File: pymc4/inference/sampling.py [hunk 1]
```diff
@@ -174,7 +187,17 @@ def logpfn(*values, **kwargs):
         _, st = flow.evaluate_model_transformed(model, state=st)
         return st.collect_log_prob()
 
-    return logpfn, dict(state.all_unobserved_values)
+    @tf.function(autograph=False)
+    def deterministics_callback(*values, **kwargs):
+        if kwargs and values:
```
Comment (ferrine) [inline pymc4/inference/sampling.py]: So what you actually want to do is a second model pass. Right? I agree it is the only solution without tfp codebase change. Looks like an acceptable solution. I hope tf is smart enough to optimize the execution graph, can you please check the slowdown?

File: tests/test_sampling.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+
+@pytest.fixture(scope="module", params=[True, False], ids=str)
+def xla_fixture(request):
+    return request.param
+
+
+def test_sample_deterministics(simple_model_with_deterministic, xla_fixture):
+    if xla_fixture:
+        pytest.skip("XLA in sampling is still not fully supported")
```
Comment (ferrine) [inline tests/test_sampling.py]: Ok,  at least w/o should work.

File: tests/test_sampling.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    @pm.model()
+    def simple_model_with_deterministic():
+        norm = yield simple_model()
+        determ = yield pm.Deterministic("determ", norm * 2)
+        return determ
+
+    return simple_model_with_deterministic
+
+
+@pytest.fixture(scope="module", params=[True, False], ids=str)
```
Comment (ferrine) [inline tests/test_sampling.py]: Better to use descriptive ids xla/noxla. True/False are out of context in test log.

File: pymc4/inference/sampling.py (hunk 1 above)
Comment (lucianopaz) [reply pymc4/inference/sampling.py]: I'll try to measure the slowdown and post here shortly

File: tests/test_executor.py [hunk 4]
```diff
@@ -442,3 +461,21 @@ def model():
...
+    assert set(expected_deterministics) <= set(state.deterministics)
+    for expected_deterministic, op, op_inputs in zip(
+        expected_deterministics, expected_ops, expected_ops_inputs
+    ):
+        inputs = [v for k, v in state.all_values.items() if k in op_inputs]
+        out = op(*inputs)
+        np.testing.assert_allclose(state.deterministics[expected_deterministic], out)
+
+
+def test_vectorize_log_prob_det_function(model_with_deterministics):
```
[the remark this answers is not in the document]
Comment (lucianopaz) [reply tests/test_executor.py]: I'm planning on testing how the deterministic and the log_prob functions handle a model that is not inherently vectorized. For example, something that uses a reduction operator, like `max`, that would squeeze all the dimensions from a tensor in a normal pass through the model, shouldn't squeeze the batch part when we use `vectorized_map`.

**Comment by lucianopaz** [comment]
I measured the slowdown on a simple model:

```python
import timeit
import pymc4 as pm


if hasattr(pm, "Deterministic"):
    @pm.model()
    def model():
        norm = yield pm.Normal("norm", 0, 1)
        determ = yield pm.Deterministic("determ", norm * 2)
        return norm
else:
    @pm.model()
    def model():
        norm = yield pm.Normal("norm", 0, 1)
        return norm


duration = timeit.timeit(lambda : pm.inference.sampling.sample(model()), number=10)
print(duration/10)
```

I got 3.9 seconds per sample call without deterministics and 4 seconds per sample call with the current implementation of deterministics. That's a 2,5% slowdown.

**Comment by lucianopaz** [comment]
Ok, I'm not sure why the tests are failing on azure. They pass without any problems locally.

File: pymc4/inference/sampling.py (hunk 1 above)
Comment (lucianopaz) [reply pymc4/inference/sampling.py]: I measured the slowdown [here](https://github.com/pymc-devs/pymc4/pull/161#issuecomment-555506333). It's around 2.5%

File: pymc4/inference/utils.py [hunk 5]
```diff
@@ -15,10 +17,14 @@ def initialize_state(model: Model, observed: Optional[dict] = None) -> flow.Samp
 
     Returns
     -------
-    pymc4.flow.SamplingState
+    state: pymc4.flow.SamplingState
+        The model's sampling state
+    deterministic_names: set
```
Comment (ferrine) [inline pymc4/inference/utils.py]: list

File: pymc4/flow/executor.py [hunk 6]
```diff
@@ -509,6 +537,26 @@ def proceed_distribution(
         state.distributions[scoped_name] = dist
         return return_value, state
 
+    def proceed_deterministic(
```
Comment (ferrine) [inline pymc4/flow/executor.py]: Here
https://github.com/pymc-devs/pymc4/blob/4685b96604290b21a16275d90d3e33ed6bc740f1/pymc4/flow/executor.py#L588

We may have an output that requested to be recorded. I would treat it as deterministic too

Comment (ferrine) [reply pymc4/flow/executor.py]: Output of the model

File: pymc4/inference/utils.py (hunk 5 above)
Comment (lucianopaz) [reply pymc4/inference/utils.py]: Oh, right!

File: pymc4/flow/executor.py (hunk 6 above)
Comment (lucianopaz) [reply pymc4/flow/executor.py]: Ok, so mean that I would replace this:
`state.untransformed_values[return_name] = return_value`
with this:
`state.deterministics[return_name] = return_value`, right?

**Comment by lucianopaz** [comment]
@ferrine, thanks for the careful reviews. I applied your comments. The later one led me to change the expected output of some tests.

**Comment by ferrine** [comment]
Cool

**Pull request merged**